### PR TITLE
[SYCL][ESIMD] Add function to get reference to underlying data

### PR DESCRIPTION
**Pull request by sarnex**
This is required for inline assembly.

### CODE REVIEW

**Comment by sarnex** [comment]
Tests: https://github.com/intel/llvm-test-suite/pull/1675

**Comment by sarnex** [comment]
Feedback from @v-klochkov 

1) Err for uses of data_ref() if not in inline asm

2) Update esimd spec

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include "TreeTransform.h"
 #include "clang/AST/AST.h"
 #include "clang/AST/Mangle.h"
+#include "clang/AST/ParentMapContext.h"
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:14]: Why was this required?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -11799,6 +11799,8 @@ def err_sycl_device_function_is_called_from_esimd : Error<
   "SYCL device function cannot be called from an ESIMD context">;
 def err_sycl_esimd_vectorize_unsupported_value : Error<
   "%0 attribute argument must be 8, 16, or 32">;
+def err_sycl_esimd_invalid_use_outline_inlineasm : Error<
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:11802]: responded to this below but tldr nothing happens and it works fine

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (sarnex) [reply clang/lib/Sema/SemaSYCL.cpp:14]: yeah for the loop

File: clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+int x = 5;
+};
+
+
+__attribute__((sycl_device)) void func() {
+ simd_obj_impl obj;
+
+// expected-error@+1 {{Function may only be used in inline assembly}}
+int& loc = obj.data_ref();
+__asm__("%0" : "=r"(obj.data_ref()));
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: Yes, there is not a correctness requirement, it will work fine with your example. However our team decided we don't want to give users a reference to the underlying vector, and we are only doing it because it's required for doing assignment in inline asm. So it's a design restriction not a technical one

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -581,12 +582,24 @@ class DiagDeviceFunction : public RecursiveASTVisitor<DiagDeviceFunction> {
...
         if (Method->isVirtual() &&
             !SemaRef.getLangOpts().SYCLAllowVirtualFunctions)
           SemaRef.Diag(e->getExprLoc(), diag::err_sycl_restrict)
               << Sema::KernelCallVirtualFunction;
 
+        QualType Ty =
+            SemaRef.getASTContext().getRecordType(Method->getParent());
+        if (isSyclType(Ty, SYCLTypeAttr::esimd_simd_or_simd_mask) &&
+            Method->getNameAsString() == "data_ref") {
+          for (auto Use : SemaRef.getASTContext().getParents(*e)) {
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Sema/SemaSYCL.cpp:595]: Glad I had you review this to catch that, thanks!

Yes, i think we need to do this in FE. Once we get to the middle end, the function call is always inlined and we can't reliably detect the bad case anymore.. Can you point me to an example or test that uses deferred diagnostics, I'm not familiar with then.

[the remark this answers is not in the document]
Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:595]: For deferred diagnostics in the source code, start with `SYCLDiagIfDeviceCode(...)` in `SemaSYCL.cpp` and you can trace what we do there.

For a test that uses it, see e.g. `SemaSYCL/sycl-restrict.cpp`.

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Sema/SemaSYCL.cpp:595]: thanks, i think i see what to do. ill update this pr once i have something

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Sema/SemaSYCL.cpp:595]: nevermind, im stuck again. i can't find anywhere where we have the expr of the function call when doing deferred diagnostics, which is what need here since we only want to allow this function call for a particular use (directly inside an asm stmt). did i miss something?

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Sema/SemaSYCL.cpp:595]: @premanandrao @elizabethandrews any ideas on the above? thanks!

File: clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: I spoke to the CFE architects about this diagnostic and there are many concerns. The design is problematic as it does not fully consider all the ways this function can be called. For example, what about calls through function pointers.  Also what stops users from using inline ASM in their own function to smuggle the reference out? A diagnostic which correctly/fully handles this is not really feasible in the frontend @erichkeane @premanandrao @AaronBallman

[the remark this answers is not in the document]
Comment (v-klochkov) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: IMO, it is still good to have such simple check.

It is like a small fence at the edge of a cliff on a hiking route fall-preventing those who walk and look at their phone, or who doesn't have idea where it stops being safe.
Surely, some still can jump over the fence, or steal the pointer returned by ref_data() via using asm, and  then hurt themselves, but that would be on their own risk/expense.
This error is a clear and visible message that the method is public, but is designed for limited usage - in asm only.

BTW, it can be a warning instead of an error as well.

[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: This sounds like a job for a static-analyzer for me.  Warnings on otherwise valid code that is guaranteed to do what they want, should not be warned on.

[the remark this answers is not in the document]
Comment (v-klochkov) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: @erichkeane - when you say "This sounds like a job for a static-analyzer for me". Do you mean this check needs to be somewhere in different place, not where it is now?

[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: Yes, a static analyzer is a separate tool from a compiler, like Klockworks or Coverity (or clang-tidy, or the Clang Static Analyzer, both of which are NOT the frontend).

[the remark this answers is not in the document]
Comment (v-klochkov) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: Static analyzers like Coverity will not find unintended uses of a public method ref_data() because it doesn't know which use is  good and which is not. Also, nobody runs Coverity-like tools for usual sycl apps.

The idea was to signal on too risky/unintended usage of the method before the code fails/hangs/etc at runtime.

If it is too expensive to do this check here, then Ok, I understand. Then, perhaps this check can be moved to some other place either in C++FE (deferred diagnostics?), or in the middle (sycl-post-link + use_def_analysis in one of passes like LowerESIMD).

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: There is no place in the FE for this diagnostic and I'd be pretty strongly opposed to diagnosing it from the backend on usability grounds if the optimizer could end up changing the diagnostic behavior (such a diagnostic would not be upstreamed to Clang/LLVM, most likely). As Erich said, there are two places in the toolchain stack where this could potentially live: either as a static analyzer check or in a clang-tidy check. I agree that no third-party static analysis tool is interested in SYCL currently, so that basically leaves the clang static analyzer (and clang-tidy) as the only choice there.

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: so i have a local implementation of an assert in the middle end (inside an ESIMD pass run in sycl-post-link) that just errors with 'unsupported use of function' or something like that. the implementation is simple, does not compromise compile time or runtime performance, and i do not expect the optimizer to cause false positives. are you all okay with doing it this way? if so i can clean it up and post it. thanks for the discussion.

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: well my understanding of @v-klochkov's point is that the function does work fine, but it's very easy for users to shoot themselves in the foot, so we wanted to add an error/warning to only allow it in tested cases. personally i don't mind dropping the error/warning, i think giving the warning in the documentation for the function is probably enough

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: I'm surprised anyone would argue that using inline assembly is *less likely* to cause a user to shoot themselves in the foot, but at the end of the day, such a design decision is very unusual for C++ and should be a red flag that the design has issues. If this is too dangerous to call directly, it should not be exposed to users. If it needs to be exposed to users, you should remove the sharp edges so they can use it. But changing the toolchain to work around a difficult-to-use design feels like the cart driving the horse.

[the remark this answers is not in the document]
Comment (sarnex) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: @v-klochkov do you have a case in mind where the user could hurt themselves calling the function directly? if so, can we do anything do fix that?

[the remark this answers is not in the document]
Comment (v-klochkov) [reply clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp:16]: Ok, let's drop the check for now.

File: sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp [hunk 5]
```diff
@@ -354,6 +354,11 @@ class simd_obj_impl {
 #endif
   }
 
+  /// @return A reference to the value of the
+  /// underlying raw vector. Intended for use
+  /// with inline assembly.
+  raw_vector_type &data_ref() { return M_data; }
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp:360]: Just idea:
Is it possible to reuse the name 'data()', similarly to std::vector::data(): https://en.cppreference.com/w/cpp/container/vector/data
or it is not doable, because the returned value of data() and data_ref() differs not only by a 'const' specifier?

**Comment by v-klochkov** [comment]
@sarnex - I think the tests from https://github.com/intel/llvm-test-suite/pull/1675 needs to be moved to this PR (sycl/test-e2e).

**Comment by sarnex** [comment]
@v-klochkov ill move the tests, thanks

File: sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp (hunk 5 above)
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/detail/simd_obj_impl.hpp:360]: no we can do that, good idea, thanks

**Review by v-klochkov — APPROVED** [review]
